### ISSUE TITLE
Not able to retrieve the some value in EDIT view

### ISSUE THREAD

**Issue by Aravin** [nathanvda/cocoon#467]
I am able to save the data using Cocoon. However, in edit view, I am not getting some values into textbox.

I know, it's not cocoon issue. but please provide me any suggestion/idea.

---
### Controller
new
`@invoice.invoice_items.build` 

create
`@invoice.save` 

-- it will save in 3 tables - invoices, invoice_items, items

edit
`@invoice = Invoice.find(params[:id])`

-- I am not able to retrieve the values from items values and show it in the textbox. Other fields are populated.

---

### Models

Invoice

```
has_many :invoice_items, inverse_of: :invoice, dependent: :destroy
```

InvoiceItem

```
belongs_to :invoice
belongs_to :item
```
Item

_nothing..._

### Output
After some workaround, I can get IDs of item table using `@invoice.invoice_items.map(&:item_id)`

If I pass the same in dropdown box, only first item gets selected.
`<% item = options_for_select(item_hash, @invoice.invoice_items.collect(&:item_id)) %>`

### In HTML

2 Times this will occurs
```
<option value="4">Item 1</option>
<option value="5" selected="selected">Item 2</option>
<option value="6">Movie Tikcet</option>
<option value="7" selected="selected">Test</option></select>
```
See item 5 and 7 both are selected becuased invoice contain that both item. 

**So, I want first item should be selected 5 and second item should be 7 and so on..**

**Comment by nathanvda** [comment]
Please clarify: 

* are you are you not able to save `items` --> or did it jus start like this and this was solved? 
* how do your associations/models look? 
* how do your views look? 

Are you able to save and just want to set the default value for `item_id` ? 

* you can build the first item in the controller and set `item_id`
* you can use a wrapper object to set the `item_id` for all further added items
* you can set the `item_id` using javascript using the callbacks on insert

**Comment by Aravin** [comment]
nathanvda, 

* I am able to save easily. Because, I will table item_id in Invoice table.
* Please find model and associations in bug description (added recently).
* Please find the view below.

--
* No. I am not able to build the first item also. I mean, I can build `quality, unit cost`, etc. since it is in Invoice model. But item `name and item description` are not building since it is in Item Model.
* I will try to add wrapper object
* Javascript is fine but if I have 10+ items. How to add to respective `item_id `to respective row.

### View

```
  <td class="item_name_desc">
    <% item_hash = @items.map {|i| [i['name'],i['id'], {'description' => i['description']}, {'cost' => i['cost']} ]} %>
    <% item = options_for_select(item_hash, @invoice.invoice_items.collect(&:item_id)) %>
    <%= f.select :item_id, item, { :include_blank => '- select -'}, {:class => 'item_name', :onChange => 'itemDropdownChange(this)' }  %>
    <%= f.text_area :item_description, class: 'item_description' %>
  </td>
  <td>
    <%= f.text_field :quantity, type: :number, required: 'required',  onChange: 'calcInvoiceTotal()' %>
  </td>
  <td>
    <%= f.text_field :unit_cost, required: "required", class: 'item_unit_cost', onChange: 'calcInvoiceTotal()' %>
  </td>
  <td>
    <%= f.text_field :total_item_cost, required: "required", class: 'item_total_cost', readonly:'readonly' %>
  </td>
```

**Comment by Aravin** [comment]
Any update @nathanvda ?

**Comment by nathanvda** [comment]
Hi @aravin I am just back from holiday, sorry if I do not reply quickly enough, for free, right. I have to work for a living too. 

Anyway: I also just saw your exact same question on stackoverflow, and I do not understand why you do not seem to use cocoon to collect the items. Imho you have `invoice -> has_many invoice_items -> has_many items` and does your views match this structure? Because you only show a single partial which is ... not part of cocoon? Please show more complete structure, e.g. 

* your edit view
* has `fields_for :invoice_items` ?
* has then a nested `fields_for :items` ? in which case I do not understand why you do this weird thing with the `item_id`? 

Or you are asking how to fill up a multi-select dropdown view? Just set the `multiple` option to true. Please check the documentation: https://apidock.com/rails/ActionView/Helpers/FormTagHelper/select_tag

**Comment by Aravin** [comment]
I can understand @nathanvda !! Sorry to disturb you!!!

You can see all my models here: https://gitlab.com/aravin/invoice-app/tree/master/app/models

Almost complete source code here: https://gitlab.com/aravin/invoice-app/tree/master/app

Nested Page: https://gitlab.com/aravin/invoice-app/blob/master/app/views/invoices/_form.html.erb  (line # 62)

Screenshot: 
![image](https://user-images.githubusercontent.com/4869265/32457317-b854e026-c34e-11e7-83f3-7e85286d1320.png)

**Comment by Mirv** [comment]
I'm still having trouble grasping the problem ...  is it that your form has the same item selected for every form item populated when you go to edit?

**Comment by Aravin** [comment]
Yeah @mirv, I want respective vales in the drop-down.

As of now, first item is getting selected.

**Comment by Mirv** [comment]
So what you're saying is you expect ...
```
id => "Name" ...

 1 => "Test", "25 litter bag"
 37 => "Some Other Test", "Testing in progress" 
```

BUT ... Right now, you are getting the result in your picture ... which is hard for us to read behind the scenes.  I'm going to guess here & it could be wrong, but you need to check it in your rails console ...

From what I can tell, "Item Name" is being loaded into the form the same for everyone (not desired).  The weird part is that "Description", "Quantity" etc all seem to be displaying unique settings.

So basically, you are just setting up the form wrong for the default selected feature (which populates everything fine, but sets the selected name/id afterwards to all rows being the same).

**Comment by nathanvda** [comment]
Hi @aravin, now this is confusing and please try to explain what you are trying to achieve, because I understand the following from your code: 

* an invoice has many invoice-items
* an invoice-item links an item to an invoice, so an invoice-item can have only one item! 

Somehow, now, you try to show all possible selected items for the complete invoice for each invoice-item, and you seem to just only want the selected item for that invoice-item to be visible, correct? 

Then I think the fix is really simple, unless I completely misunderstand you, you should just write: 

```
    <%= f.collection_select(:item_id, @items, :id, :name) %>
```

I am using the shorter version for readability purposes, this will automatically build the correct list of options and set the value of `item_id` as selected. You seem to be building a more complicated hash I guess for js --in the `onChange`. 

So to remain closer to your code, imho you should just write:

```
    <%= f.select :item_id, options_for_select(item_hash), { :include_blank => '- select -'}, {:class => 'item_name', :onChange => 'itemDropdownChange(this)' }  %>
```

**Comment by Aravin** [comment]
Dear @nathanvda bro,

Thanks a lot bro!!

this line
```
    <%= f.collection_select(:item_id, @items, :id, :name) %>

```
 solved my issue :)

**Comment by Aravin** [comment]
@Mirv , thanks for your suggestion and SO links, its really helpful.

@nathanvda, as you mentioned, I am trying to build hash for javascript functionality. `<%= f.collection_select(:item_id, @items, :id, :name) %>` is working but I got another problem i.e description is not getting selected.

**Comment by Aravin** [comment]
Dear @nathanvda & @Mirv,

I have completed.

This is my final code,
```
<% item_hash = @items.map {|i| [i['name'],i['id'], {'description' => i['description']}, {'cost' => i['cost']} ]} %>
    <%= f.select :item_id, item_hash, { :include_blank => '- select -'}, {:class => 'item_name', :onChange => 'itemDropdownChange(this)' }  %>
    <%= f.text_area :item_description, class: 'item_description' %>
```

You can see demo app here: https://invoice-app-testing.herokuapp.com/invoices/2/edit

**Comment by Aravin** [comment]
@nathanvda regarding your comment https://github.com/nathanvda/cocoon/issues/467#issuecomment-342327978,

I realized my mistake and modified the model.

```
class Invoice < ApplicationRecord
  has_many :taxes
  has_many :payments
  has_one :contact

  belongs_to :contact

  has_many :invoice_items, inverse_of: :invoice, dependent: :destroy
  has_many :items, :through => :invoice_items, :class_name => 'Item'

  accepts_nested_attributes_for :items
  accepts_nested_attributes_for :invoice_items, :allow_destroy => :true, update_only: :true

   end
```

Thanks, bro :)
